File: ffi-client/examples/triangle_headless.rs
//! Headless triangle example using goldy-ffi-client Scheme render pass + grant readback.
//!
//! Mirrors the dotnet/python headless triangle smoke tests.
//!
//! Run from `goldy/ffi-client`: `cargo run --example triangle_headless`

use goldy_ffi_client::{
    shader::builtins, BufferKind, Color, Context, DepthFormat, DeviceDescriptor, Instance, NodeAccess, RenderPipeline,
    RenderPipelineDesc, RequestAdapterOptions, RetainedPool, Scheme, ShaderModule, TargetLoad, TextureFlags,
    TextureFormat, TextureKind, Vertex2D,
};

fn main() -> goldy_ffi_client::Result<()> {
    println!("Goldy triangle_headless (ffi-client / Scheme)\n");

    let instance = Instance::new()?;
    let device = instance
        .request_adapter(&RequestAdapterOptions::default())?
        .request_device(&DeviceDescriptor::default())?;
    let ctx = Context::new(&device)?;

    let vertices = [
        Vertex2D {
            position: [0.0, -0.5],
            color: [1.0, 0.0, 0.0, 1.0],
        },
        Vertex2D {
            position: [-0.5, 0.5],
            color: [0.0, 1.0, 0.0, 1.0],
        },
        Vertex2D {
            position: [0.5, 0.5],
            color: [0.0, 0.0, 1.0, 1.0],
        },
    ];
    let mut retained_pool = RetainedPool::new(&device)?;
    let vertex_buffer = retained_pool.acquire_buffer_with_data(&vertices, BufferKind::Scattered)?;

    const WIDTH: u32 = 64;
    const HEIGHT: u32 = 64;
    let readback = retained_pool.acquire_texture(
        WIDTH,
        HEIGHT,
        TextureFormat::Rgba8Unorm,
        TextureKind::Direct,
        TextureFlags::COPY_SRC.union(TextureFlags::COPY_DST),
        None,
    )?;

    let shader = ShaderModule::from_slang(&device, builtins::VERTEX_COLOR_2D)?;
    let pipeline = RenderPipeline::new(
        &device,
        &shader,
        &shader,
        &RenderPipelineDesc {
            vertex_layout: Vertex2D::layout(),
            target_format: TextureFormat::Rgba8Unorm,
            ..Default::default()
        },
    )?;

    let mut scheme = Scheme::new(&ctx)?;
    let rt = scheme.lease_render_target(WIDTH, HEIGHT, TextureFormat::Rgba8Unorm, None::<DepthFormat>)?;
    {
        let mut pass = scheme.render_pass("triangle", &rt, TargetLoad::Clear(Color::BLACK));
        pass.with_buffer(&vertex_buffer, NodeAccess::Read);
        pass.set_pipeline(&pipeline);
        pass.set_vertex_buffer(0, &vertex_buffer);
        pass.draw(0..3, 0..1);
        pass.finish_recorded();
    }
    scheme.copy_to_texture(&rt, &readback)?;
    let grant = scheme.grant_read_texture(&readback)?;
    let submission = scheme.submit()?;
    let pixels = grant.consume(&submission)?;
    assert!(pixels.iter().any(|&b| b > 0), "readback should contain lit pixels");

    println!("Triangle rendered and read back successfully ({} bytes).", pixels.len());
    Ok(())
}

File: dotnet/Goldy.Examples/TriangleHeadless.cs
using System.Runtime.InteropServices;
using Goldy;

/// <summary>
/// Headless triangle via retained Scheme (render pass + readback).
/// </summary>
static class TriangleHeadless
{
    [StructLayout(LayoutKind.Sequential)]
    struct Vertex2D
    {
        public float Px, Py;
        public float R, G, B, A;
    }

    public static void Run()
    {
        Console.WriteLine("Goldy .NET Triangle (headless Scheme)");
        Console.WriteLine(new string('=', 40));

        using var instance = new Instance();
        using var device = instance.RequestAdapter().RequestDevice();
        using var ctx = device.CreateContext();
        Console.WriteLine($"Backend: {instance.BackendType}");

        using var shader = new ShaderModule(device, ShaderModule.BuiltinVertexColor2D);
        using var pipeline = new RenderPipeline(
            device,
            shader,
            shader,
            new RenderPipelineDesc
            {
                VertexAttributes = VertexLayouts.Vertex2D,
                VertexStride = 24,
                TargetFormat = TextureFormat.Rgba8Unorm,
            });

        ReadOnlySpan<Vertex2D> vertices =
        [
            new() { Px = 0.0f, Py = -0.5f, R = 1, G = 0, B = 0, A = 1 },
            new() { Px = -0.5f, Py = 0.5f, R = 0, G = 1, B = 0, A = 1 },
            new() { Px = 0.5f, Py = 0.5f, R = 0, G = 0, B = 1, A = 1 },
        ];
        using var retainedPool = new RetainedPool(device);
        using var vertexParcel = retainedPool.AcquireBuffer(vertices, BufferKind.Scattered);
        using var readback = retainedPool.AcquireTexture(
            100,
            100,
            TextureFormat.Rgba8Unorm,
            TextureKind.Direct,
            TextureFlags.CopySrc | TextureFlags.CopyDst);

        const uint width = 100;
        const uint height = 100;

        using var scheme = new Scheme(ctx);
        using var rt = scheme.LeaseRenderTarget(width, height, TextureFormat.Rgba8Unorm);
        using (var pass = scheme.RenderPass("triangle", rt))
        {
            pass
                .WithParcel(vertexParcel, NodeAccess.Read)
                .Clear(new Color(0.1f, 0.1f, 0.2f, 1.0f))
                .SetPipeline(pipeline)
                .SetVertexBuffer(0, vertexParcel)
                .Draw(3);
        }

        scheme.CopyToTexture(rt, readback);
        using var grant = scheme.GrantReadTexture(readback);
        using var submission = scheme.Submit();
        var pixels = grant.Consume(submission);

        if (pixels.Length != width * height * 4)
            throw new InvalidOperationException($"Unexpected readback size: {pixels.Length}");

        var hasColor = false;
        for (var i = 0; i < pixels.Length; i += 4)
        {
            if (pixels[i] > 0 || pixels[i + 1] > 0 || pixels[i + 2] > 0)
            {
                hasColor = true;
                break;
            }
        }

        if (!hasColor)
            throw new InvalidOperationException("Triangle should write non-black pixels");

        Console.WriteLine($"Rendered {width}x{height}, readback OK");
        Console.WriteLine("Done!");
    }
}
